### PR TITLE
Include check for logsafe Preconditions

### PR DESCRIPTION
**Pull request by cbrockington**
Safe loggable Preconditions should also (if not more so) have a constant message since the message is intended to be safely loggable.

https://github.com/palantir/safe-logging/pull/28

### CODE REVIEW

**Comment by gracew** [comment]
can you add a test case for this? let's also add `com.palantir.logsafe.Preconditions` to the checkstyle exclusion for AvoidStaticImport

**Comment by cbrockington** [comment]
Just updating on this, added tests and realised there are some more changes needed as the RegexpSingleline check for "do you have {} on the same line as the word Preconditions" is no longer satisfactory, since Guava Preconditions and LogSafe Preconditions use different formatting.

Fixing that up, will let you know when it's done.

**Comment by cbrockington** [comment]
OK, this should now work properly.  Most of the changes are splitting out the RegexpSinglelineJava Preconditions formatting check into two checks: `GuavaPreconditionsMessageFormat` and `LogSafePreconditionsMessageFormat`, since a simple regex can not determine the difference between an imported guava Preconditions vs logsafe Preconditions (and they have opposite formatting expectations.

File: .baseline/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -296,11 +296,6 @@
             <property name="format" value="\? extends Object\W"/>
             <property name="message" value="Use ? rather than ? extends Object."/>
         </module>
-        <module name="RegexpSinglelineJava">
```
Comment (uschi2000) [inline .baseline/checkstyle/checkstyle.xml:299]: I don't think we should remove this. It catches many bugs that vanilla Guava users produce.

Comment (uschi2000) [reply .baseline/checkstyle/checkstyle.xml:299]: oh, you replaced it with error-prone. got it.

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/GuavaPreconditionsMessageFormat.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Override
+    protected Description matchMessageFormat(MethodInvocationTree tree, String message, VisitorState state) {
+        if (!message.contains("{}")) {
+            return Description.NO_MATCH;
+        }
+
+        return buildDescription(tree).setMessage(
+                "Use printf-style formatting in Preconditions, not '{}' style formatting.").build();
```
Comment (uschi2000) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/GuavaPreconditionsMessageFormat.java]: "in Guava Preconditions"

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LogSafePreconditionsMessageFormat.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.regex.Pattern;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "LogSafePreconditionsMessageFormat",
+        category = Category.ONE_OFF,
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = LinkType.CUSTOM,
+        severity = SeverityLevel.ERROR,
+        summary = "logsafe Preconditions.checkX() methods must have '{}' style formatting.")
```
Comment (uschi2000) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LogSafePreconditionsMessageFormat.java]: @gracew why do we accept parameters at all?

Comment (iamdanfox) [reply baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LogSafePreconditionsMessageFormat.java]: Seems like params are just more helpful information that devs can use to try to figure out what went wrong when a stack trace appears in their low-side logs...

why would we not want them?

Comment (uschi2000) [reply baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LogSafePreconditionsMessageFormat.java]: we don't inline parameters for slf4j log calls, so we would we support them here?

Comment (iamdanfox) [reply baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LogSafePreconditionsMessageFormat.java]: Oh right, you mean just getting rid of the `{}`... yeah that seems fine to me!  As long as people can keep tacking on values to the varargs call and they end up on disk somehow :)

Comment (cbrockington) [reply baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LogSafePreconditionsMessageFormat.java]: The logsafe preconditions doesn't inline them either, but I do think that even if they aren't actually substituted they can be useful (particularly when named) as they are a visual indicator "you should look in the params block if this is important to you", without them you don't necessarily know if a noun is just a noun, or a placeholder for a param.

Comment (cbrockington) [reply baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LogSafePreconditionsMessageFormat.java]: OK, I changed the check to merely say what it was actually doing anyway, which is flagging you should not be using print-f style formatting, regardless of using any braces.

Is that fair?

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LogSafePreconditionsMessageFormat.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    protected Description matchMessageFormat(MethodInvocationTree tree, String message, VisitorState state) {
+        if (!message.contains("%s")) {
+            return Description.NO_MATCH;
+        }
+
+        return buildDescription(tree).setMessage(
+                "Do not use printf-style formatting in logsafe Preconditions.").build();
```
Comment (iamdanfox) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LogSafePreconditionsMessageFormat.java:63]: ...use further `SafeArg` or `UnsafeArg` arguments instead?

**Review by uschi2000 — APPROVED** [review]

**Comment by uschi2000** [comment]
cheers, @cbrockington

**Pull request merged**